dos/src/main.rs

```diff
@@ -414,34 +418,43 @@ fn run_dos_transactions<T: 'static + BenchTpsClient + Send + Sync>(
 
     let mut transaction_generator = TransactionGenerator::new(transaction_params);
 
+    let connection_cache_stats = Arc::new(ConnectionCacheStats::default());
+    let udp_client = UdpTpuConnection::new(target, connection_cache_stats);
+
     let mut count = 0;
     let mut total_count = 0;
     let mut error_count = 0;
     let mut last_log = Instant::now();
+
+    let mut data = Vec::<Vec<u8>>::with_capacity(SEND_BATCH_MAX_SIZE);
     loop {
-        let chunk_keypairs = if generate_keypairs {
-            let permutation = it.next();
-            if permutation.is_none() {
-                // if ran out of permutations, regenerate keys
-                keypairs_flat.iter_mut().for_each(|v| *v = Keypair::new());
-                info!("Regenerate keypairs");
-                continue;
-            }
-            let permutation = permutation.unwrap();
-            Some(apply_permutation(permutation, &keypairs_flat))
-        } else {
-            None
-        };
+        let send_batch_size = min(iterations - total_count, SEND_BATCH_MAX_SIZE);
+        data.clear();
+        for _ in 0..send_batch_size {
+            let chunk_keypairs = if generate_keypairs {
+                let mut permutation = it.next();
+                if permutation.is_none() {
+                    // if ran out of permutations, regenerate keys
+                    keypairs_flat.iter_mut().for_each(|v| *v = Keypair::new());
+                    info!("Regenerate keypairs");
+                    permutation = it.next();
+                }
+                let permutation = permutation.unwrap();
+                Some(apply_permutation(permutation, &keypairs_flat))
+            } else {
+                None
+            };
+            let tx = transaction_generator.generate(payer, chunk_keypairs, client.as_ref());
+            data.push(bincode::serialize(&tx).unwrap());
+        }
 
-        let tx = transaction_generator.generate(payer, chunk_keypairs, client.as_ref());
+        let res = udp_client.send_wire_transaction_batch(&data);
```

Comment:
untested idea: you can consider using `send_wire_transaction_batch_async` instead to really pack in messages using QUIC! It won't make any difference for UDP

Reply:
should be better with async, I agree